src/DynamoCore/DSEngine/NodeToCode.cs

```diff
@@ -1,16 +1,847 @@
+                    // variables are renamed to this one because of renumbering.
+                    // If there is confliction, continue to bump its ID.
+                    ns = new NumberingState(ident);
+                    numberingMap[ident] = ns;
+
+                    while (mappedVariables.Contains(ns.NumberedVariable))
+                        ns.BumpID();
+                }
+
+                if (ns.ID != 0)
+                {
+                    n.Name = n.Value = ns.NumberedVariable;
+
+                    // If this variable is numbered, and it is also output to
+                    // other node, we should remap the output variable to
+                    // this re-numbered variable.
+                    string name = string.Format("{0}%{1}", ident, node.GUID);
+                    if (renamingMap.ContainsKey(name))
+                    {
+                        var mappedName = renamingMap[name];
+                        renamingMap[mappedName] = ns.NumberedVariable;
+                    }
+
+                    // Record in output map.
+                    if (outputMap.ContainsKey(name))
+                    {
+                        var mappedName = outputMap[name];
+                        outputMap[mappedName] = ns.NumberedVariable;
+                    }
+                }
+
+                mappedVariables.Add(ns.NumberedVariable);
+            };
+
+            IdentifierVisitor visitor = new IdentifierVisitor(func);
+            astNode.Accept(visitor);
+        }
+
+        /// <summary>
+        /// Remap variables.
+        /// </summary>
+        /// <param name="astNode"></param>
+        /// <param name="renamingMap"></param>
+        private static void VariableRemapping(
+            AssociativeNode astNode, 
+            Dictionary<string, string> renamingMap)
+        {
+            Action<IdentifierNode> func = n =>
+                {
+                    string newIdent;
+                    if (renamingMap.TryGetValue(n.Value, out newIdent))
+                        n.Value = n.Name = newIdent;
+                };
+
+            IdentifierVisitor visitor = new IdentifierVisitor(func);
+            astNode.Accept(visitor);
+        }
+
+        /// <summary>
+        /// Map variable to shorter name.
+        /// </summary>
+        /// <param name="astNode"></param>
+        /// <param name="shortNameMap"></param>
+        /// <param name="nameGenerator"></param>
+        private static void ShortNameMapping(
+            AssociativeNode astNode,
+            Dictionary<string, string> shortNameMap,
+            ShortNameGenerator nameGenerator,
+            HashSet<string> mappedVariables)
+        {
+            Action<IdentifierNode> func = n =>
+            {
+                string shortName;
+                if (shortNameMap.TryGetValue(n.Value, out shortName))
+                {
+                    if (shortName == string.Empty)
```

Comment:
OK. So `shortNameMap` is either `inputMap` or `outputMap`. Either, they could have empty string value, which means we haven't generated a short name for some input key yet.